# 第一章：探索性数据分析

## 载入数据

### 任务一：导入numpy和pandas

In [1]:
import numpy as np
import pandas as pd

### 任务二：载入数据

(1) 使用相对路径载入数据   

In [2]:
df = pd.read_csv("train.csv")

In [3]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

(2) 使用绝对路径载入数据

In [4]:
df2 = pd.read_csv("C:\\Users\\Administrator\\Desktop\\hands-on-data-analysis-master\\第一单元项目集合\\train.csv")

In [5]:
df2.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### 任务三：每1000行为一个数据模块，逐块读取

什么是逐块读取？  
https://blog.csdn.net/sinat_29957455/article/details/79057650

In [6]:
chunker = pd.read_csv("train.csv",chunksize=1000)

In [7]:
for piece in chunker:
    print(type(piece))
    print(len(piece))

<class 'pandas.core.frame.DataFrame'>
891


### 任务四：将表头改成中文，索引改为乘客ID

In [11]:
df = df.rename(columns={"PassengerId":"乘客ID","Survived":"是否幸存",
                  "Pclass":"乘客等级(1/2/3等舱位) ","Name":"乘客姓名",
                  "Sex":"性别","Age":"年龄","SibSp":"堂兄弟/妹个数",
                  "Parch":"父母与小孩个数","Ticket":"船票信息","Fare":"票价",
                  "Cabin":"客舱","Embarked":"登船港口"})

In [12]:
df.head()

乘客ID  是否幸存  乘客等级(1/2/3等舱位)   \
0     1     0                3   
1     2     1                1   
2     3     1                3   
3     4     1                1   
4     5     0                3   

                                                乘客姓名      性别    年龄  堂兄弟/妹个数  \
0                            Braund, Mr. Owen Harris    male  22.0        1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0        1   
2                             Heikkinen, Miss. Laina  female  26.0        0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0        1   
4                           Allen, Mr. William Henry    male  35.0        0   

   父母与小孩个数              船票信息       票价    客舱 登船港口  
0        0         A/5 21171   7.2500   NaN    S  
1        0          PC 17599  71.2833   C85    C  
2        0  STON/O2. 3101282   7.9250   NaN    S  
3        0            113803  53.1000  C123    S  
4        0            373450   8.0500   NaN    S

In [13]:
# 方法二
df1 = pd.read_csv('train.csv', names=['乘客ID','是否幸存','仓位等级','姓名','性别','年龄',
                                      '兄弟姐妹个数','父母子女个数','船票信息','票价','客舱','登船港口'],
                  index_col='乘客ID',header=0)

In [14]:
df.head()

乘客ID  是否幸存  乘客等级(1/2/3等舱位)   \
0     1     0                3   
1     2     1                1   
2     3     1                3   
3     4     1                1   
4     5     0                3   

                                                乘客姓名      性别    年龄  堂兄弟/妹个数  \
0                            Braund, Mr. Owen Harris    male  22.0        1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0        1   
2                             Heikkinen, Miss. Laina  female  26.0        0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0        1   
4                           Allen, Mr. William Henry    male  35.0        0   

   父母与小孩个数              船票信息       票价    客舱 登船港口  
0        0         A/5 21171   7.2500   NaN    S  
1        0          PC 17599  71.2833   C85    C  
2        0  STON/O2. 3101282   7.9250   NaN    S  
3        0            113803  53.1000  C123    S  
4        0            373450   8.0500   NaN    S

## 初步观察

导入数据后，你可能要对数据的整体结构和样例进行概览，比如说，数据大小、有多少列，各列都是什么格式的，是否包含null等

### 任务一：查看数据的基本信息

In [20]:
df.values  # 查看所有元素

array([[1, 0, 3, ..., 7.25, nan, 'S'],
       [2, 1, 1, ..., 71.2833, 'C85', 'C'],
       [3, 1, 3, ..., 7.925, nan, 'S'],
       ...,
       [889, 0, 3, ..., 23.45, nan, 'S'],
       [890, 1, 1, ..., 30.0, 'C148', 'C'],
       [891, 0, 3, ..., 7.75, nan, 'Q']], dtype=object)

In [21]:
df.columns # 查看所有列名

Index(['乘客ID', '是否幸存', '乘客等级(1/2/3等舱位) ', '乘客姓名', '性别', '年龄', '堂兄弟/妹个数',
       '父母与小孩个数', '船票信息', '票价', '客舱', '登船港口'],
      dtype='object')

In [26]:
df.size    # 元素总数

10692

In [27]:
df.ndim    # 表的维度数

2

In [28]:
df.shape   # 返回表的行数与列数

(891, 12)

In [19]:
# DataFrame的详细内容
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   乘客ID             891 non-null    int64  
 1   是否幸存             891 non-null    int64  
 2   乘客等级(1/2/3等舱位)   891 non-null    int64  
 3   乘客姓名             891 non-null    object 
 4   性别               891 non-null    object 
 5   年龄               714 non-null    float64
 6   堂兄弟/妹个数          891 non-null    int64  
 7   父母与小孩个数          891 non-null    int64  
 8   船票信息             891 non-null    object 
 9   票价               891 non-null    float64
 10  客舱               204 non-null    object 
 11  登船港口             889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [17]:
df.describe()

乘客ID        是否幸存  乘客等级(1/2/3等舱位)           年龄     堂兄弟/妹个数  \
count  891.000000  891.000000       891.000000  714.000000  891.000000   
mean   446.000000    0.383838         2.308642   29.699118    0.523008   
std    257.353842    0.486592         0.836071   14.526497    1.102743   
min      1.000000    0.000000         1.000000    0.420000    0.000000   
25%    223.500000    0.000000         2.000000   20.125000    0.000000   
50%    446.000000    0.000000         3.000000   28.000000    0.000000   
75%    668.500000    1.000000         3.000000   38.000000    1.000000   
max    891.000000    1.000000         3.000000   80.000000    8.000000   

          父母与小孩个数          票价  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

### 任务二：观察表格前10行的数据和后15行的数据

In [18]:
# 前10行数据
df.head(10)

乘客ID  是否幸存  乘客等级(1/2/3等舱位)   \
0     1     0                3   
1     2     1                1   
2     3     1                3   
3     4     1                1   
4     5     0                3   
5     6     0                3   
6     7     0                1   
7     8     0                3   
8     9     1                3   
9    10     1                2   

                                                乘客姓名      性别    年龄  堂兄弟/妹个数  \
0                            Braund, Mr. Owen Harris    male  22.0        1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0        1   
2                             Heikkinen, Miss. Laina  female  26.0        0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0        1   
4                           Allen, Mr. William Henry    male  35.0        0   
5                                   Moran, Mr. James    male   NaN        0   
6                            McCarthy, Mr. Timothy J    male  54.0        0   
7                     Palsson, Master. Gosta Leonard    male   2.0        3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0        0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0        1   

   父母与小孩个数              船票信息       票价    客舱 登船港口  
0        0         A/5 21171   7.2500   NaN    S  
1        0          PC 17599  71.2833   C85    C  
2        0  STON/O2. 3101282   7.9250   NaN    S  
3        0            113803  53.1000  C123    S  
4        0            373450   8.0500   NaN    S  
5        0            330877   8.4583   NaN    Q  
6        0             17463  51.8625   E46    S  
7        1            349909  21.0750   NaN    S  
8        2            347742  11.1333   NaN    S  
9        0            237736  30.0708   NaN    C

In [29]:
# 后15行数据
df.tail(15)

乘客ID  是否幸存  乘客等级(1/2/3等舱位)   \
876   877     0                3   
877   878     0                3   
878   879     0                3   
879   880     1                1   
880   881     1                2   
881   882     0                3   
882   883     0                3   
883   884     0                2   
884   885     0                3   
885   886     0                3   
886   887     0                2   
887   888     1                1   
888   889     0                3   
889   890     1                1   
890   891     0                3   

                                              乘客姓名      性别    年龄  堂兄弟/妹个数  \
876                  Gustafsson, Mr. Alfred Ossian    male  20.0        0   
877                           Petroff, Mr. Nedelio    male  19.0        0   
878                             Laleff, Mr. Kristo    male   NaN        0   
879  Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0        0   
880   Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0        0   
881                             Markun, Mr. Johann    male  33.0        0   
882                   Dahlberg, Miss. Gerda Ulrika  female  22.0        0   
883                  Banfield, Mr. Frederick James    male  28.0        0   
884                         Sutehall, Mr. Henry Jr    male  25.0        0   
885           Rice, Mrs. William (Margaret Norton)  female  39.0        0   
886                          Montvila, Rev. Juozas    male  27.0        0   
887                   Graham, Miss. Margaret Edith  female  19.0        0   
888       Johnston, Miss. Catherine Helen "Carrie"  female   NaN        1   
889                          Behr, Mr. Karl Howell    male  26.0        0   
890                            Dooley, Mr. Patrick    male  32.0        0   

     父母与小孩个数              船票信息       票价    客舱 登船港口  
876        0              7534   9.8458   NaN    S  
877        0            349212   7.8958   NaN    S  
878        0            349217   7.8958   NaN    S  
879        1             11767  83.1583   C50    C  
880        1            230433  26.0000   NaN    S  
881        0            349257   7.8958   NaN    S  
882        0              7552  10.5167   NaN    S  
883        0  C.A./SOTON 34068  10.5000   NaN    S  
884        0   SOTON/OQ 392076   7.0500   NaN    S  
885        5            382652  29.1250   NaN    Q  
886        0            211536  13.0000   NaN    S  
887        0            112053  30.0000   B42    S  
888        2        W./C. 6607  23.4500   NaN    S  
889        0            111369  30.0000  C148    C  
890        0            370376   7.7500   NaN    Q

### 任务三：判断数据是否为空，为空的地方返回True，其余地方返回False

In [31]:
df.isna().head()

乘客ID   是否幸存  乘客等级(1/2/3等舱位)    乘客姓名     性别     年龄  堂兄弟/妹个数  父母与小孩个数  \
0  False  False            False  False  False  False    False    False   
1  False  False            False  False  False  False    False    False   
2  False  False            False  False  False  False    False    False   
3  False  False            False  False  False  False    False    False   
4  False  False            False  False  False  False    False    False   

    船票信息     票价     客舱   登船港口  
0  False  False   True  False  
1  False  False  False  False  
2  False  False   True  False  
3  False  False  False  False  
4  False  False   True  False

In [32]:
df.isnull().head()

乘客ID   是否幸存  乘客等级(1/2/3等舱位)    乘客姓名     性别     年龄  堂兄弟/妹个数  父母与小孩个数  \
0  False  False            False  False  False  False    False    False   
1  False  False            False  False  False  False    False    False   
2  False  False            False  False  False  False    False    False   
3  False  False            False  False  False  False    False    False   
4  False  False            False  False  False  False    False    False   

    船票信息     票价     客舱   登船港口  
0  False  False   True  False  
1  False  False  False  False  
2  False  False   True  False  
3  False  False  False  False  
4  False  False   True  False

## 保存数据

### 任务一：将你加载并做出改变的数据，在工作目录下保存为一个新文件train_chinese.csv

In [33]:
df.to_csv('train_chinese.csv')

## 知道你的数据叫什么

我们学习pandas的基础操作，那么上一节通过pandas加载之后的数据，其数据类型是什么呢？

### 任务一：pandas中有两个数据类型DateFrame和Series，通过查找简单了解他们。

In [37]:
data1 = {"a":10,"b":20,"c":30,"d":40}
pd.Series(data1)

a    10
b    20
c    30
d    40
dtype: int64

In [38]:
data2 = {'state': ['Ohio', 'Ohio', 'Ohio', 'Nevada', 'Nevada', 'Nevada'],
        'year': [2000, 2001, 2002, 2001, 2002, 2003],'pop': [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}
pd.DataFrame(data2)

state  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

### 任务二：根据上节课的方法载入"train.csv"文件

In [41]:
df1 = pd.read_csv("train.csv")

In [42]:
df1.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### 任务三：查看DataFrame数据的每列的项

In [43]:
df1.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### 任务四：查看"cabin"这列的所有项

In [46]:
df1["Cabin"].head()

0     NaN
1     C85
2     NaN
3    C123
4     NaN
Name: Cabin, dtype: object

In [48]:
df1.Cabin.head()

0     NaN
1     C85
2     NaN
3    C123
4     NaN
Name: Cabin, dtype: object

### 任务五：加载文件"test_1.csv"，然后对比"train.csv"，看看有哪些多出的列，然后将多出的列删除

In [57]:
df2 = pd.read_csv("test_1.csv", index_col=0)
df2.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked    a  
0      0         A/5 21171   7.2500   NaN        S  100  
1      0          PC 17599  71.2833   C85        C  100  
2      0  STON/O2. 3101282   7.9250   NaN        S  100  
3      0            113803  53.1000  C123        S  100  
4      0            373450   8.0500   NaN        S  100

In [51]:
df1.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [58]:
del df2['a']

In [59]:
df2.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

【思考】还有其他的删除多余的列的方式吗？

In [63]:
df2.drop(["Embarked"],1).head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin  
0      0         A/5 21171   7.2500   NaN  
1      0          PC 17599  71.2833   C85  
2      0  STON/O2. 3101282   7.9250   NaN  
3      0            113803  53.1000  C123  
4      0            373450   8.0500   NaN

### 任务六： 将['PassengerId','Name','Age','Ticket']这几个列元素隐藏，只观察其他几个列元素

In [64]:
df1.drop(["PassengerId","Name","Age","Ticket"],1).head()

Survived  Pclass     Sex  SibSp  Parch     Fare Cabin Embarked
0         0       3    male      1      0   7.2500   NaN        S
1         1       1  female      1      0  71.2833   C85        C
2         1       3  female      0      0   7.9250   NaN        S
3         1       1  female      1      0  53.1000  C123        S
4         0       3    male      0      0   8.0500   NaN        S

【思考】对比任务五和任务六，是不是使用了不一样的方法(函数)，如果使用一样的函数如何完成上面的不同的要求呢？

In [65]:
# del删除多列报错
df1 = pd.read_csv("train.csv")
df1.del(["PassengerId","Name","Age","Ticket"])

SyntaxError: invalid syntax (<ipython-input-65-17d3feace5b2>, line 2)

【思考回答】
如果想要完全的删除你的数据结构，使用inplace=True，因为使用inplace就将原数据覆盖了，所以这里没有用

## 筛选的逻辑

表格数据中，最重要的一个功能就是要具有可筛选的能力，选出我所需要的信息，丢弃无用的信息。

### 任务一： 我们以"Age"为筛选条件，显示年龄在10岁以下的乘客信息。

In [66]:
df1.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [68]:
df1[df1["Age"]<10]

PassengerId  Survived  Pclass                                      Name  \
7              8         0       3            Palsson, Master. Gosta Leonard   
10            11         1       3           Sandstrom, Miss. Marguerite Rut   
16            17         0       3                      Rice, Master. Eugene   
24            25         0       3             Palsson, Miss. Torborg Danira   
43            44         1       2  Laroche, Miss. Simonne Marie Anne Andree   
..           ...       ...     ...                                       ...   
827          828         1       2                     Mallet, Master. Andre   
831          832         1       2           Richards, Master. George Sibley   
850          851         0       3   Andersson, Master. Sigvard Harald Elias   
852          853         0       3                   Boulos, Miss. Nourelain   
869          870         1       3           Johnson, Master. Harold Theodor   

        Sex   Age  SibSp  Parch           Ticket     Fare Cabin Embarked  
7      male  2.00      3      1           349909  21.0750   NaN        S  
10   female  4.00      1      1          PP 9549  16.7000    G6        S  
16     male  2.00      4      1           382652  29.1250   NaN        Q  
24   female  8.00      3      1           349909  21.0750   NaN        S  
43   female  3.00      1      2    SC/Paris 2123  41.5792   NaN        C  
..      ...   ...    ...    ...              ...      ...   ...      ...  
827    male  1.00      0      2  S.C./PARIS 2079  37.0042   NaN        C  
831    male  0.83      1      1            29106  18.7500   NaN        S  
850    male  4.00      4      2           347082  31.2750   NaN        S  
852  female  9.00      1      1             2678  15.2458   NaN        C  
869    male  4.00      1      1           347742  11.1333   NaN        S  

[62 rows x 12 columns]

### 任务二： 以"Age"为条件，将年龄在10岁以上和50岁以下的乘客信息显示出来，并将这个数据命名为midage

In [72]:
midage = df1[(df1["Age"]>10) & (df1["Age"]<50)]

In [73]:
midage

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
885      5            382652  29.1250   NaN        Q  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[576 rows x 12 columns]

【提示】了解pandas的条件筛选方式以及如何使用交集和并集操作

### 任务三：将midage的数据中第100行的"Pclass"和"Sex"的数据显示出来

In [84]:
midage = midage.reset_index(drop=True)

In [85]:
# 重置索引
midage

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
571          886         0       3   
572          887         0       2   
573          888         1       1   
574          890         1       1   
575          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
571               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
572                              Montvila, Rev. Juozas    male  27.0      0   
573                       Graham, Miss. Margaret Edith  female  19.0      0   
574                              Behr, Mr. Karl Howell    male  26.0      0   
575                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
571      5            382652  29.1250   NaN        Q  
572      0            211536  13.0000   NaN        S  
573      0            112053  30.0000   B42        S  
574      0            111369  30.0000  C148        C  
575      0            370376   7.7500   NaN        Q  

[576 rows x 12 columns]

【思考】这个reset_index()函数的作用是什么？如果不用这个函数，下面的任务会出现什么情况？

In [86]:
midage.loc[[100],['Pclass','Sex']]

Pclass   Sex
100       2  male

### 任务四：将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来

In [88]:
midage.loc[[100,105,108],['Pclass','Name','Sex']]

Pclass                               Name   Sex
100       2  Byles, Rev. Thomas Roussel Davids  male
105       3           Cribb, Mr. John Hatfield  male
108       3                    Calic, Mr. Jovo  male

【提示】使用pandas提出的简单方式，你可以看看loc方法  
对比整体的数据位置，你有发现什么问题吗？那么如何解决？

### 任务五：使用iloc方法将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来

In [90]:
midage.iloc[[100,105,108],[2,3,4]]

Pclass                               Name   Sex
100       2  Byles, Rev. Thomas Roussel Davids  male
105       3           Cribb, Mr. John Hatfield  male
108       3                    Calic, Mr. Jovo  male

## 了解你的数据吗？

### 任务一：利用Pandas对示例数据进行排序，要求升序

In [92]:
test = pd.DataFrame(np.arange(8).reshape((2, 4)),
                   index=[2, 1],
                   columns=['d', 'a', 'b', 'c']) 

In [93]:
test

d  a  b  c
2  0  1  2  3
1  4  5  6  7

【问题】：大多数时候我们都是想根据列的值来排序,所以将你构建的DataFrame中的数据根据某一列，升序排列

In [94]:
test.sort_values(by='a', ascending=False)

d  a  b  c
1  4  5  6  7
2  0  1  2  3

【思考】通过书本你能说出Pandas对DataFrame数据的其他排序方式吗？  
【总结】下面将不同的排序方式做一个总结

1.让行索引升序排序

In [96]:
test.sort_index(axis=0, ascending=True) #默认升序

d  a  b  c
1  4  5  6  7
2  0  1  2  3

2.让列索引升序排序

In [97]:
test.sort_index(axis=1)

a  b  c  d
2  1  2  3  0
1  5  6  7  4

3.让列索引降序排序

In [98]:
test.sort_index(axis=1,ascending=False)

d  c  b  a
2  0  3  2  1
1  4  7  6  5

4.让任选两列数据同时降序排序

In [102]:
test.sort_values(by=['a','c'],ascending=False)

d  a  b  c
1  4  5  6  7
2  0  1  2  3

### 任务二：对泰坦尼克号数据（trian.csv）按票价和年龄两列进行综合排序（降序排列），从这个数据中你可以分析出什么？

In [104]:
df = pd.read_csv("train.csv")

In [105]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [114]:
df.sort_values(by=['Fare','Age'],ascending=False).head()

PassengerId  Survived  Pclass                                Name  \
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
258          259         1       1                    Ward, Miss. Anna   
737          738         1       1              Lesurer, Mr. Gustave J   
438          439         0       1                   Fortune, Mr. Mark   
341          342         1       1      Fortune, Miss. Alice Elizabeth   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C  
258  female  35.0      0      0  PC 17755  512.3292          NaN        C  
737    male  35.0      0      0  PC 17755  512.3292         B101        C  
438    male  64.0      1      4     19950  263.0000  C23 C25 C27        S  
341  female  24.0      3      2     19950  263.0000  C23 C25 C27        S

【思考】排序后，如果我们仅仅关注年龄和票价两列。根据常识我知道发现票价越高的应该客舱越好，所以我们会明显看出，票价前20的乘客中存活的有14人，这是相当高的一个比例，那么我们后面是不是可以进一步分析一下票价和存活之间的关系，年龄和存活之间的关系呢？当你开始发现数据之间的关系了，数据分析就开始了。

多做几个数据的排序

In [118]:
df.sort_values(by=['Age','Survived'],ascending=False).head(20)

PassengerId  Survived  Pclass                                       Name  \
630          631         1       1       Barkworth, Mr. Algernon Henry Wilson   
851          852         0       3                        Svensson, Mr. Johan   
96            97         0       1                  Goldschmidt, Mr. George B   
493          494         0       1                    Artagaveytia, Mr. Ramon   
116          117         0       3                       Connors, Mr. Patrick   
672          673         0       2                Mitchell, Mr. Henry Michael   
745          746         0       1               Crosby, Capt. Edward Gifford   
33            34         0       2                      Wheadon, Mr. Edward H   
54            55         0       1             Ostby, Mr. Engelhart Cornelius   
280          281         0       3                           Duane, Mr. Frank   
456          457         0       1                  Millet, Mr. Francis Davis   
438          439         0       1                          Fortune, Mr. Mark   
545          546         0       1               Nicholson, Mr. Arthur Ernest   
275          276         1       1          Andrews, Miss. Kornelia Theodosia   
483          484         1       3                     Turkula, Mrs. (Hedwig)   
570          571         1       2                         Harris, Mr. George   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   
252          253         0       1                  Stead, Mr. William Thomas   
555          556         0       1                         Wright, Mr. George   
170          171         0       1                  Van der hoef, Mr. Wyckoff   

        Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked  
630    male  80.0      0      0        27042   30.0000          A23        S  
851    male  74.0      0      0       347060    7.7750          NaN        S  
96     male  71.0      0      0     PC 17754   34.6542           A5        C  
493    male  71.0      0      0     PC 17609   49.5042          NaN        C  
116    male  70.5      0      0       370369    7.7500          NaN        Q  
672    male  70.0      0      0   C.A. 24580   10.5000          NaN        S  
745    male  70.0      1      1    WE/P 5735   71.0000          B22        S  
33     male  66.0      0      0   C.A. 24579   10.5000          NaN        S  
54     male  65.0      0      1       113509   61.9792          B30        C  
280    male  65.0      0      0       336439    7.7500          NaN        Q  
456    male  65.0      0      0        13509   26.5500          E38        S  
438    male  64.0      1      4        19950  263.0000  C23 C25 C27        S  
545    male  64.0      0      0          693   26.0000          NaN        S  
275  female  63.0      1      0        13502   77.9583           D7        S  
483  female  63.0      0      0         4134    9.5875          NaN        S  
570    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S  
829  female  62.0      0      0       113572   80.0000          B28      NaN  
252    male  62.0      0      0       113514   26.5500          C87        S  
555    male  62.0      0      0       113807   26.5500          NaN        S  
170    male  61.0      0      0       111240   33.5000          B19        S

高龄60岁以上存活率很低

### 任务三：利用Pandas进行算术计算，计算两个DataFrame数据相加结果

In [119]:
frame1_a = pd.DataFrame(np.arange(9.).reshape(3, 3),
                     columns=['a', 'b', 'c'],
                     index=['one', 'two', 'three'])
frame1_b = pd.DataFrame(np.arange(12.).reshape(4, 3),
                     columns=['a', 'e', 'c'],
                     index=['first', 'one', 'two', 'second'])

In [121]:
frame1_a

a    b    c
one    0.0  1.0  2.0
two    3.0  4.0  5.0
three  6.0  7.0  8.0

In [122]:
frame1_b

a     e     c
first   0.0   1.0   2.0
one     3.0   4.0   5.0
two     6.0   7.0   8.0
second  9.0  10.0  11.0

将frame_a和frame_b进行相加

In [123]:
frame1_a + frame1_b

a   b     c   e
first   NaN NaN   NaN NaN
one     3.0 NaN   7.0 NaN
second  NaN NaN   NaN NaN
three   NaN NaN   NaN NaN
two     9.0 NaN  13.0 NaN

【提醒】两个DataFrame相加后，会返回一个新的DataFrame，对应的行和列的值会相加，没有对应的会变成空值NaN。
当然，DataFrame还有很多算术运算，如减法，除法等，有兴趣的同学可以看《利用Python进行数据分析》第五章 算术运算与数据对齐 部分，多在网络上查找相关学习资料。

### 任务四：通过泰坦尼克号数据如何计算出在船上最大的家族有多少人？

In [127]:
max(df['SibSp']+df['Parch'])

10

【提醒】我们只需找出”兄弟姐妹个数“和”父母子女个数“之和最大的数，当然你还可以想出很多方法和思考角度

In [131]:
df['family']=df['SibSp']+df['Parch']
df.sort_values(by='family',ascending=False).head()

PassengerId  Survived  Pclass                          Name     Sex  Age  \
180          181         0       3  Sage, Miss. Constance Gladys  female  NaN   
846          847         0       3      Sage, Mr. Douglas Bullen    male  NaN   
792          793         0       3       Sage, Miss. Stella Anna  female  NaN   
201          202         0       3           Sage, Mr. Frederick    male  NaN   
159          160         0       3    Sage, Master. Thomas Henry    male  NaN   

     SibSp  Parch    Ticket   Fare Cabin Embarked  family  
180      8      2  CA. 2343  69.55   NaN        S      10  
846      8      2  CA. 2343  69.55   NaN        S      10  
792      8      2  CA. 2343  69.55   NaN        S      10  
201      8      2  CA. 2343  69.55   NaN        S      10  
159      8      2  CA. 2343  69.55   NaN        S      10

### 任务五：学会使用Pandas describe()函数查看数据基本统计信息

In [132]:
frame2 = pd.DataFrame([[1.4, np.nan], 
                       [7.1, -4.5],
                       [np.nan, np.nan], 
                       [0.75, -1.3]
                      ], index=['a', 'b', 'c', 'd'], columns=['one', 'two'])

In [133]:
frame2

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

In [134]:
frame2.describe()

one       two
count  3.000000  2.000000
mean   3.083333 -2.900000
std    3.493685  2.262742
min    0.750000 -4.500000
25%    1.075000 -3.700000
50%    1.400000 -2.900000
75%    4.250000 -2.100000
max    7.100000 -1.300000

### 任务六：分别看看泰坦尼克号数据集中 票价、父母子女 这列数据的基本统计数据，你能发现什么？

In [136]:
df['Fare'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

一共有891个票价数据， 平均值约为：32.20， 标准差约为49.69，说明票价波动特别大， 25%的人的票价是低于7.91的，50%的人的票价低于14.45，75%的人的票价低于31.00， 票价最大值约为512.33，最小值为0。

In [137]:
df['Parch'].describe()

count    891.000000
mean       0.381594
std        0.806057
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        6.000000
Name: Parch, dtype: float64

船上有老人和孩子的人数较少，平均值为0.38，75%的人都不是老人和孩子。

In [138]:
df['Age'].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

船上共有714人知道年龄，这714个人的平均年龄为30岁，其中75%的人年龄在38岁以下

【总结】本节中我们通过Pandas的一些内置函数对数据进行了初步统计查看，这个过程最重要的不是大家得掌握这些函数，而是看懂从这些函数出来的数据，构建自己的数据分析思维，这也是第一章最重要的点，希望大家学完第一章能对数据有个基本认识，了解自己在做什么，为什么这么做，后面的章节我们将开始对数据进行清洗，进一步分析。